cf-custom-resources/lib/custom-domain-app-runner.js

```diff
@@ -0,0 +1,189 @@
+    const data = await appRunnerClient.associateCustomDomain({
+        DomainName: customDomainName,
+        ServiceArn: serviceARN,
+    }).promise().catch(err => {
+        if (err.name === ERR_NAME_INVALID_REQUEST && err.message.includes(`${customDomainName} is already associated with`)) {
+            throw new CustomDomainError(`${customDomainName} is already associated with service ${serviceARN}`, ERR_NAME_CUSTOM_DOMAIN_ALREADY_ASSOCIATED);
+        }
+        throw err;
+    });
+
+    await upsertCNAMERecordAndWait(customDomainName, data.DNSTarget, appHostedZoneID);
+    await validateCertForDomain(serviceARN, customDomainName);
+}
+
+/**
+ * Validate certificates of the custom domain for the service by upserting validation records.
+ *
+ * @param {string} serviceARN ARN of the service that the custom domain applies to.
+ * @param {string} domainName the custom domain name.
+ * @throws wrapped error.
+ */
+async function validateCertForDomain(serviceARN, domainName) {
+    console.log("Add validation records");
+    let i;
+    for (i = 0; i < ATTEMPTS; i++){
+        const data = await appRunnerClient.describeCustomDomains({
+            ServiceArn: serviceARN,
+        }).promise().catch(err => {
+            throw new Error(`get custom domains for service ${serviceARN}: ` + err.message);
+        });
+
+        const customDomains = data.CustomDomains;
+        let domain;
+        for (const i in customDomains) {
+            if (customDomains[i].DomainName === domainName) {
+                domain = customDomains[i];
+                break;
+            }
+        }
+        if (!domain) {
+            throw new Error(`domain ${domainName} is not associated`);
+        }
+
+        if (domain.Status !== DOMAIN_STATUS_PENDING_VERIFICATION) {
+            console.log(`Custom domain ${domainName} status is ${domain.Status}. Desired state is ${DOMAIN_STATUS_PENDING_VERIFICATION}. Wait and check again`);
+            await sleep(3000);
+            continue;
+        }
+
+        const records = domain.CertificateValidationRecords;
+        for (const i in records) {
+            await upsertCNAMERecordAndWait(records[i].Name, records[i].Value, appHostedZoneID).catch(err => {
+                throw new Error("upsert certificate validation record: " + err.message);
+            });
+        }
+        break;
+    }
+
+    if (i >= ATTEMPTS) {
+        throw new Error(`failed waiting for custom domain ${domainName} to change to state ${DOMAIN_STATUS_PENDING_VERIFICATION}`);
+    }
+}
+
+/**
+ * Upserts a CNAME record and wait for the change to have taken place.
+ *
+ * @param {string} recordName the name of the record
+ * @param {string} recordValue the value of the record
+ * @param {string} hostedZoneID the ID of the hosted zone into which the record needs to be upserted.
+ * @throws wrapped error.
+ */
+async function upsertCNAMERecordAndWait(recordName, recordValue, hostedZoneID) {
```

Comment:
```suggestion
async function updateCNAMERecordAndWait(recordName, recordValue, hostedZoneID, action) {
```
there might be some duplicated code once we add `delete` action handling.